### PR TITLE
ADD/COPY command analysis by tern

### PR DESCRIPTION
**Pull request by mukultaneja**
Recognize ADD/COPY command by Dockerfile/
created_by parser

This change notifies about not able to
analyze few commands such as ADD/COPY
by tern in the generated report.

Work towards: #929

Signed-off-by: Mukul Taneja <mtaneja@vmware.com>

### CODE REVIEW

**Comment by mukultaneja** [comment]
@nishakm, Kindly review the updated changes, and I will update the same PR with Dockerfile parser changes as well

**Comment by mukultaneja** [comment]
@nishakm 
I ran below Dockerfile and tried to see if the PR changes worked fine on it. I got the below Layer info as part of the report and seems the changes worked fine.

`Command:- tern report -d Dockerfile`

==== Dockerfile ====
FROM node: slim
WORKDIR /restapi
ADD requirements.txt /restapi
ADD install.txt /restapi

Report = 
Layer 5:
		info: Layer created by commands: /bin/sh -c #(nop) COPY file:238737301d47304174e4d24f4def935b29b3069c03c72ae8de97d94624382fce in /usr/local/bin/
		warning: unknown file. please analyze separately #(nop) COPY file:238737301d47304174e4d24f4def935b29b3069c03c72ae8de97d94624382fce in /usr/local/bin/

Layer 6:
		info: Layer created by commands: /bin/sh -c #(nop) WORKDIR /restapi
		warning:
Unrecognized Commands:#(nop) WORKDIR /restapi

Layer 7:
		info: Layer created by commands: /bin/sh -c #(nop) ADD file:c4b063782b7e30618696529b7ca33d29068a9b4795eebe5ba302ed419c59dd8e in /restapi
		warning: unknown file. please analyze separately #(nop) ADD file:c4b063782b7e30618696529b7ca33d29068a9b4795eebe5ba302ed419c59dd8e in /restapi

Layer 8:
		info: Layer created by commands: /bin/sh -c #(nop) ADD file:9c5bbc9ae0f138728453d2db2e387c847280db4efab7acbe48da8a95c5b96b2a in /restapi
		warning: unknown file. please analyze separately #(nop) ADD file:9c5bbc9ae0f138728453d2db2e387c847280db4efab7acbe48da8a95c5b96b2a in /restapi

**Comment by nishakm** [comment]
@mukultaneja the COPY instruction is not recognized when running `tern lock Dockerfile`. There's a roundabout way you can test this by using an existing go project. I've used https://github.com/nishakm/genv because it's simple.
1. Clone this repo
2. cd into the repo
3. Run `go build genv.go`. It should produce a binary called `genv`.
4. make the following Dockerfile:
```
FROM gcr.io/distroless/base-debian10
COPY genv /
CMD ["/genv --help"]
```
5. Run `tern lock Dockerfile`. This will generate a file called `Dockerfile.lock`
For me, this file looks like this:
```
FROM gcr.io/distroless/base-debian10@sha256:6ec6da1888b18dd971802c2a58a76a7702902b4c9c1be28f38e75e871cedc2df
COPY genv /
CMD ["/genv --help"]
```
We expect some information about the github commit and project URL here.

To see what that looks like for the ADD command, create another Dockerfile that looks like this:
```
FROM golang:1.13-buster
WORKDIR /go/src/app
ADD . /go/src/app
RUN go get -d -v ./...
RUN go build -o /go/bin/app
CMD ["/go/bin/app/genv --help"]
```
Then run `tern lock Dockerfile`. I get something like this:
```
FROM golang@sha256:66a3f6817c129f48c9cc9b96816a1c0b6dcb399393196e3581d9367b55ab29f2
WORKDIR /go/src/app
ADD . /go/src/app # git project name: genv, HEAD sha: ec701242124523627c64f5ada5ef946ba12d87b3, project url(s): git@github.com:nishakm/genv.git
RUN go get -d -v ./...
RUN go build -o /go/bin/app
CMD ["/go/bin/app/genv --help"]
```
Notice that `ADD` has the git project name, commit and url. We want the same thing for `COPY`.
```

**Comment by nishakm** [comment]
@mukultaneja You will just need to add `COPY` to ` def expand_add_command(dfobj):` in `parse.py`.

File: tern/report/errors.py [hunk 1]
```diff
@@ -81,6 +81,7 @@
     '''registry.\n'''
 empty_layer = '''Empty layer. Nothing to do.\n'''
 no_layer_created_by = "No created_by information for layer."
+no_able_to_analyze = "unknown {entity}. please analyze separately "
```
Comment (rnjudge) [inline tern/report/errors.py:84]: small nit: `not_able_to_analyze` and capitalize first words of each sentence + add period: `Unknown {entity}. Please analyze separately."

**Comment by mukultaneja** [comment]
@nishakm @rnjudge, how about if we change the `expand_add_command` method name to `expand_commands` as we are having two commands to expand in this method now. Further, we can have more if needed.

**Comment by nishakm** [comment]
> @nishakm @rnjudge, how about if we change the `expand_add_command` method name to `expand_commands` as we are having two commands to expand in this method now. Further, we can have more if needed.

@rnjudge is better at naming than I am. I would rather it be renamed as `expand_add_copy_command` if you care to rename it.

**Comment by mukultaneja** [comment]
> @mukultaneja Would you be able to rebase your commit on top of the main branch? That way, the tests will include the change to fix the "lock" test.

@nishakm yeah, I will do it today.

**Comment by mukultaneja** [comment]
@nishakm, I rebased my branch. please review once and merge the changes if all looks good!

**Pull request merged**